lib/filewatch/since_db.rb

```diff
@@ -0,0 +1,247 @@
+          # or a different file with the same content
+          # leave as is
+        end
+        @logger.debug? && @logger.debug("SinceDb find: returning value")
+        return value
+      end
+
+      old_key = new_key = nil
+
+      @short_keys.each do |k, vk|
+        next if k >= key.size
+        @logger.debug? && @logger.debug("SinceDb find: key: #{key}, short_keys - #{k}:#{vk}")
+        # vk is a list of keys at size k
+        old_key, new_key = wf.first_fingerprint_match_any?(vk.sort)
+        # old_key is the one found via short_keys
+        # new_key is the altered wf.storage key
+        # it would have a new shorter fingerprint
+        break if old_key && new_key # can't set during iteration
+      end
+      if old_key && new_key && member?(old_key)
+        @logger.debug? && @logger.debug("SinceDb find: short keys match - #{old_key}, #{new_key}")
+        value = delete(old_key)
+        if value.watched_file.nil?
+          @logger.debug? && @logger.debug("SinceDb find: short keys unallocated wf")
+          wf.update_bytes_read(value.position)
+          value.upd_watched_file(wf)
+        elsif value.watched_file != wf
+          @logger.debug? && @logger.debug("SinceDb find: short keys allocated wf unequal")
+          # put back
+          set(old_key, value)
+          # new value
+          value = SincedbValue.new(0) # allow the new allocation to start from scratch
+          value.upd_watched_file(wf)
+        else
+          @logger.debug? && @logger.debug("SinceDb find: short keys allocated wf equal")
+          # its the same watched file with a new key
+          # nothing to do
+        end
+        set(new_key, value)
+        @logger.debug? && @logger.debug("SinceDb find: found! - #{value}")
+        return value
+      end
+      @logger.debug? && @logger.debug("SinceDb find: NOT FOUND!, path: #{wf.path}")
+      nil
+    end
+
+    def member?(key)
+      @sincedb.member?(key)
+    end
+
+    def move(k1, k2)
+      set(k2, delete(k1))
+    end
+
+    def get(key)
+      @sincedb[key]
+    end
+
+    def delete(key)
+      @sincedb.delete(key)
+    end
+
+    def last_read(key)
+      @sincedb[key].position
+    end
+
+    def rewind(key)
+      @sincedb[key].upd_position(0)
+    end
+
+    def store_last_read(key, last_read)
+      @sincedb[key].upd_position(last_read)
+    end
+
+    def increment(key, amount)
+      @sincedb[key].inc_position(amount)
+    end
+
+    def add_watched_file(key, wf)
+      @sincedb[key].upd_watched_file(wf)
+    end
+
+    def clear
+      @sincedb.clear
+    end
+
+    def keys
+      @sincedb.keys
+    end
+
+    def set(key, value)
+      @sincedb[key] = value
+      if key.version?(2) && key.short?
+        @short_keys[key.size] << key
+      end
+      value
+    end
+
+    def unallocated?(key)
+      return false if !member?(key)
+      get(key).watched_file.nil?
+    end
+
+    private
+
+    def store_from_file(line)
```

Comment:
Well, its storing the record in the DB, but in the spirit of ETL i guess `load` is a POLS metaphor.